### PR TITLE
fix(cpp) fix class/template parsing issues

### LINKED ISSUE #2102: (cpp) grammar confused by complex Template << >> patterns
This code results in everything after `L':'` to be highlighted as if it was a string: **(Editor: This is now resolved, the second issue still standing).**

```cpp
template<typename T>
class PropertyChangedBase {
public:
    winrt::event_token PropertyChanged(const winrt::Windows::UI::Xaml::Data::PropertyChangedEventHandler &value)
    {
        return m_propertyChanged.add(value);
    }

    void PropertyChanged(const winrt::event_token &token)
    {
        m_propertyChanged.remove(token);
    }

protected:
    template<typename U>
    void compare_assign(U &value, const U &new_value, std::wstring_view name)
    {
        if (value != new_value)
        {
            value = new_value;
            name.remove_prefix(name.find_last_of(L':') + 1);
            m_propertyChanged(*static_cast<T *>(this), winrt::Windows::UI::Xaml::Data::PropertyChangedEventArgs(name));
        }
    }

private:
    winrt::event<winrt::Windows::UI::Xaml::Data::PropertyChangedEventHandler> m_propertyChanged;
};
```

On this one, it just completely stops highlighting after the second `is_streamable`

```cpp
namespace impl {
    template<typename T, typename = void>
    struct is_streamable : std::false_type { };

    template<typename T>
    struct is_streamable<T, std::void_t<decltype(std::declval<std::wostream &>() << std::declval<T>())>> : std::true_type { };

    template<typename T>
    inline constexpr bool is_streamable_v = is_streamable<T>::value;
}

// Disable overload for already valid operands.
template<class T, class = std::enable_if_t<!impl::is_streamable_v<const T &> && std::is_convertible_v<const T &, std::wstring_view>>>
std::wostream &operator <<(std::wostream &stream, const T &thing)
{
    return stream << static_cast<std::wstring_view>(thing);
}
```

Both pieces above is completely correct code which compiles, assuming the right headers are included

### LINKED ISSUE #2716: (cpp) Template parameters trigger class header rule
**Describe the issue**

When using the `template <class T>` syntax in C++, the `class` keyword misleads the parser into thinking it’s found a class header, and goes into class title mode, if there’s some `{` shortly afterwards. The relevant rule is

https://github.com/highlightjs/highlight.js/blob/5b87cc4d106059bee1c30f123e5a4053e0af9de4/src/languages/c-like.js#L227

**Which language seems to have the issue?**

`cpp`

**Are you using `highlight` or `highlightAuto`?**

`highlight`. When using `highlightAuto`, the code snippet is mis-categorised as `angelscript` and parsed differently still.

**Sample Code to Reproduce**

```cpp
template <class T> // comment
auto foo(T x) { … }
```

highlight.js marks this up as

```xml
<keyword>template</keyword> &lt;<class><keyword>class</keyword> <title>T</title>&gt; // <title>comment</title>
<title>auto</title> <title>foo</title>(<title>T</title> <title>x</title>) {</class> … }
```

**Expected behavior**

When replacing `class` with `typename` (which is 100% equivalent in this context), it’s marked up (more) correctly:

```xml
<keyword>template</keyword> &lt;<keyword>typename</keyword> T&gt; <comment>// comment</comment>
<function><keyword>auto</keyword> <title>foo</title><params>(T x)</params> </function>{ … }
```

**Additional context**

Present in master. Some earlier version used highlighted this correctly (presumably before support for class headers was added). Fundamentally the issue is that template parameter declarations aren’t handled at all by the lexer (by contrast, template argument lists *do* have rudimentary handling; i.e. usage rather than declaration).

A fix would probably require a separate mode for template parameter declarations, which is a hairy topic. In fact, correctly disambiguating the meaning of `>>` (can be right-shift operator *or* a terminating nested template argument list) is context-sensitive, and undecidable for the lexer (!) because its meaning can depend on the type of names it is applied to: `a<b<c>>d` can either be parsed as a variable declaration of `d` where the type has a nested template argument (list if `a` is a type), or as an expression involving less-than comparison and bit shifts, if `a` is a variable: https://godbolt.org/z/rM18oc

### PR DESCRIPTION
**Pull request by joshgoebel**
This fixes #2716 and it fixes #2102.

This solves the problem simply by aborting the class mode whenever we see a `<`, `>` or `=`, and returning to the regular rule matching.  This is a much simpler solution that trying to process deeply nested templates and count subexpressions of `<` vs `(` - all the while trying to figure out what is perhaps simple a less than (or greater than) vs a template delimiter.

This also adds support for `enum struct` and `enum class`.  This actually mostly worked before but now the "class" block extends around the "enum" as well which is slightly more correct I think.

- Also adds a minor linting fix to quite a few tests regarding carriage return at the end of file.

See tests.

@klmr

### CODE REVIEW

File: src/languages/c-like.js [hunk 1]
```diff
@@ -225,10 +225,9 @@ export default function(hljs) {
       },
       {
         className: 'class',
-        beginKeywords: 'class struct', end: /[{;:]/,
+        beginKeywords: 'enum class struct', end: /[{;:<>=]/,
```
Comment (klmr) [inline src/languages/c-like.js:228]: This keyword list is still missing `union`.

File: src/languages/c-like.js [hunk 2]
```diff
@@ -225,10 +225,9 @@ export default function(hljs) {
...
         contains: [
-          { beginKeywords: "final" },
-          {begin: /</, end: />/, contains: ['self']}, // skip generic stuff
+          { beginKeywords: "final class struct enum" },
```
Comment (klmr) [inline src/languages/c-like.js:230]: And I don’t think `enum` is required here.

File: test/markup/cpp/template_complexity.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+// Disable overload for already valid operands.
+template<class T, class = std::enable_if_t<!impl::is_streamable_v<const T &> && std::is_convertible_v<const T &, std::wstring_view>>>
+std::wostream &operator <<(std::wostream &stream, const T &thing)
+{
+    return stream << static_cast<std::wstring_view>(thing);
+}
+
+enum struct DataHolder { }
```
Comment (klmr) [inline test/markup/cpp/template_complexity.txt:16]: It doesn’t really matter either way but we should probably add semicolons behind this and the following type definitions to make the test cases valid C++.

**Review by klmr — CHANGES_REQUESTED** [review]
This still doesn’t handle attributes, e.g.

```cpp
struct [[deprecated]] foobar {};
```

Of course a simple fix would be to add `|\[\[` to `end` but handling them properly isn’t actually much harder: we might just need a nested rule `{begin: /\[\[/, end: /]]/, keywords: 'using'}`.

But I wouldn’t increase the scope of this PR, as such attributes still relatively rare in C++ code (and even rarer on type declarations).

File: src/languages/c-like.js (hunk 2 above)
Comment (joshgoebel) [reply src/languages/c-like.js:230]: Yeah, but doesn't hurt either. :-)

**Comment by joshgoebel** [comment]
@klmr  Better?

**Review by klmr — APPROVED** [review]

**Pull request merged**